### PR TITLE
[AIRFLOW-1998] Implemented DatabricksRunNowOperator for jobs/run-now …

### PR DESCRIPTION
**Pull request by isknight**
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AIRFLOW-1998

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

  - Added a new Databrick's operator, DatabricksRunNowOperator, for the API's jobs/run-now functionality.
  - After consulting Databricks on their preference, pulled shared code out of DatabricksSubmitRunOperator into module level functions that both the DatabricksSubmitRunOperator & DatabricksRunNowOperator call.
  - Added required API call to databricks_hook.py
  - Added/updated tests for both databricks_operators.py and the databricks_hook.py
  - Fixed a connection state issue with tokens in the setUp part of the operator tests where conn.extras wasn't getting reset between test runs and maintaining its mock token.
  - Fixed some existing indentation for flake8

### Tests

- [x] My PR adds the following unit tests:

- I mirrored the same tests that the existing DatabricksSubmitRunOperator had for the new DatabricksRunNowOperator.

  - **Added**:
tests/contrib/operators/test_databricks_operator.py
	* DatabricksOperatorSharedFunctions
		* test_deep_string_coerce
	* DatabricksRunNowOperatorTest
		* test_init_with_named_parameters
		* test_init_with_json
		* test_init_with_merging
		* test_init_with_templating
		* test_init_with_bad_type
		* test_exec_success
		* test_exec_failure
		* test_on_kill
  - **Updated**:
tests/contrib/operators/test_databricks_operator.py
	* DatabricksSubmitRunOperatorTest
		* test_init_with_named_parameters
		* test_init_with_json
		* test_init_with_merging
		* test_init_with_templating
		* test_init_with_bad_type
		* test_exec_success
		* test_exec_failure
		* test_on_kill
  - **Removed**:
tests/contrib/operators/test_databricks_operator.py
	* DatabricksSubmitRunOperatorTest
		* test_deep_string_coerce (it was moved to its own external test class)
		
		
### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/contrib/hooks/databricks_hook.py [hunk 1]
```diff
@@ -143,6 +144,18 @@ def _do_api_call(self, endpoint_info, json):
         raise AirflowException(('API requests to Databricks failed {} times. ' +
                                'Giving up.').format(self.retry_limit))
 
+    def run_now(self, json):
+        """
+        Utility function to call the ``api/2.0/jobs/run-now`` endpoint.
+
+        :param json: The data used in the body of the request to the ``submit`` endpoint.
```
Comment (andrewmchen) [inline airflow/contrib/hooks/databricks_hook.py]: nit: `submit` -> `run_now`.

Comment (andrewmchen) [inline airflow/contrib/hooks/databricks_hook.py]: nit: `submit` -> `run-now`

File: airflow/contrib/operators/databricks_operator.py [hunk 2]
```diff
@@ -30,6 +30,81 @@
...
+        return {k: c(v, '{0}[{1}]'.format(json_path, k))
+                for k, v in list(content.items())}
+    else:
+        param_type = type(content)
+        msg = 'Type {0} used for parameter {1} is not a number or a string' \
+            .format(param_type, json_path)
+        raise AirflowException(msg)
+
+
+def _handle_databricks_operator_execution(operator, context):
```
Comment (andrewmchen) [inline airflow/contrib/operators/databricks_operator.py]: Can we explicitly pass the parameters: `hook` and `log`, into this function and lift the logic to initially invoke `run-now` or `run-submit` back into the operator? The end result would be a helper function responsible for polling for the run result.

With this solution, we'd duplicate a bit of logic with the xcom business (which we can refactor later) but we avoid having to pass the wide interface of an `operator` (which is really a special databricks operato) to this function.

File: airflow/contrib/operators/databricks_operator.py [hunk 3]
```diff
@@ -30,6 +30,81 @@
...
+                    s=run_state)
+                raise AirflowException(error_message)
+        else:
+            operator.log.info('%s in run state: %s', operator.task_id, run_state)
+            operator.log_run_page_url(run_page_url)
+            operator.log.info('Sleeping for %s seconds.', operator.polling_period_seconds)
+            time.sleep(operator.polling_period_seconds)
+
+
+def _handle_databricks_operator_on_kill(operator):
```
Comment (andrewmchen) [inline airflow/contrib/operators/databricks_operator.py]: For this one, I'm thinking we can simply duplicate the logic in both operators.

**Review by andrewmchen — CHANGES_REQUESTED** [review]
Looks very good overall. I just had a couple comments about passing the "operator" to the module functions.

File: airflow/contrib/operators/databricks_operator.py (hunk 3 above)
Comment (isknight) [reply airflow/contrib/operators/databricks_operator.py]: Yeah, I felt iffy when I pulled that out. It's not good to duplicate code, but what if their lifecycle deviates? I'll update this.

File: airflow/contrib/hooks/databricks_hook.py (hunk 1 above)
Comment (isknight) [reply airflow/contrib/hooks/databricks_hook.py]: Oops! nice catch. I read through my new docs like 5 times but managed to miss it. :) I'll update.

**Comment by Fokko** [comment]
@isknight Can you rebase onto master? There are merge conflicts :(

**Comment by isknight** [comment]
Apologies @Fokko I was offline for a bit. For some odd reason my git client is erroring out when I try to interactively rebase and squash my commits. I'll look into it some more tomorrow. Otherwise, if @andrewmchen (and others) are ok with my changes, perhaps it would be simpler for me to make a new PR?

**Comment by dalupus** [comment]
lgtm

**Comment by Fokko** [comment]
No problem @isknight 

@andrewmchen any final thoughts?

File: airflow/contrib/operators/databricks_operator.py [hunk 4]
```diff
@@ -30,6 +30,66 @@
...
+    else:
+        param_type = type(content)
+        msg = 'Type {0} used for parameter {1} is not a number or a string' \
+            .format(param_type, json_path)
+        raise AirflowException(msg)
+
+
+def _handle_databricks_operator_execution(operator, hook, log, context):
+    """
+    Handles the Airflow + Databricks lifecycle logic for a data bricks operator
```
Comment (andrewmchen) [inline airflow/contrib/operators/databricks_operator.py]: nit: `Databricks`

**Comment by andrewmchen** [comment]
LGTM, thanks for the PR again!

**Review by andrewmchen — APPROVED** [review]

**Pull request merged**